### DISCUSSION TITLE
Integrating Switchbot S1

### DISCUSSION THREAD

**Discussion by 1technophile** [theengs/decoder#90] [New devices/things integration]
As a prior work to the integration of Switchbot into OMG I would like to integrate the broadcast reading of switch bot actuator, I have found this doc:
https://github.com/OpenWonderLabs/python-host/wiki/Bot-BLE-open-API#broadcast-message-format
By the way I salute the fact that SwitchBot publish its BLE API and message format 👏

From it, we can easily identify the device type with the corresponding servicedata first byte.
```
   "model_id":"SWITCHBOT-S1",
   "condition":["uuid", "index",0, "0d00","&", "servicedata", "=", 6, "index", 0, "48"],
```

Nevertheless, the other data are available in binary format, so we would need to target a byte convert it to binary, target one or several bit of this binary to compare it with some reference.
Example: `48d0db`
```
      "mode":{
         "condition":["servicedata", "bin", 2, 7, "1"],
         "decoder":["static_value", "on/off"]
      },
      "mode":{
         "condition":["servicedata", "bin", 2, 7, "0"],
         "decoder":["static_value", "one state"]
      },
      "state":{
         "condition":["servicedata", "bin", 2, 6, "1"],
         "decoder":["static_value", "on"]
      },
      "state":{
         "condition":["servicedata", "bin", 2, 6, "0"],
         "decoder":["static_value", "off"]
      }
```
2 = byte position in the servicedata (counted in char)
6 or 7 = bit position
"0" or "1" = position to check

For the battery, we may need also to extract a part of the bits and convert those to decimal, here is some bytes 2 examples:
Byte| bin| 7 bits| dec %
-- | -- | -- | --
64 | 1100100 | 1100100 | 100
5b | 1011011 | 1011011 | 91
db | 11011011 | 1011011 | 91

Example:
```
      "batt":{
         "decoder":["b_value_from_hex_data", "servicedata", 4, 0, 7]
      }
```
4 = byte position in the servicedata (counted in char)
0 = start bit to extract
7 = number of bits to extract
`48d0db`
-> `db`
-> 1011011
-> 91

@h2zero any thoughts about adding this kind of functions, other proposals are welcome?

**Comment by DigiH** [comment]
Hi @1technophile with just having created a test decoder for 

https://github.com/theengs/decoder/issues/92

also with getting not yet available binary bit data form multiple OR hex index evaluations, wouldn't this also be possible here, especially as it looks like that the SWITCHBOT-S1 has a lot less states to check ;)

Not really as a permanent solution for necessarily merging into the release, as I also don't see that for the Govee H5055 decoder in the test state, but for checking and testing the functionality for these devices until a nicer solution is available.

Thanks

**Reply by 1technophile** [reply]
Hi @DigiH 

I agree we can integrate the decoders following your proposal and optimize later on, I'm going to share some ideas on the issue for the H5055

**Reply by DigiH** [reply]
I would assume that a similar solution for the Switchbot would be a lot smaller and compact, fron the info you gave above.

**Reply by DigiH** [reply]
Hi @1technophile, while waiting to see what we are going to do with the H5055 decoder condition I have tried a quick test for your Switchbot S1 proposal here, until a direct bit-wise condition is possible.

Do you actually have this device for testing, and was the sample data `"48d0db"` for its **mode** being `"on/off"` and the **state** `"off"`? If so, I think I got it together right and you might want to try

https://github.com/DigiH/decoder/tree/switchbot-test

Thanks

**Reply by DigiH** [reply]
UPDATE: Actually just submitted a correction. From your example above 

```
      "state":{
         "condition":["servicedata", "bin", 2, 6, "1"],
         "decoder":["static_value", "on"]
      },
      "state":{
         "condition":["servicedata", "bin", 2, 6, "0"],
         "decoder":["static_value", "off"]
      }
```

I also thought it actually makes sense, "1" - ON, "0" - OFF, but in the linked API docs it's actually

```
Bit[6] - State

    0: on
    1: off
```

which could be just a typo, but easily recognisable when testing ;)

**Reply by 1technophile** [reply]
Works great thanks!
```json
{"id":"FF","rssi":-80,"servicedatauuid":"0x0d00","brand":"Switchbot","model":"S1","model_id":"SWITCHBOT-S1","mode":"on/off","state":"off","batt":91}
```

```json
{"id":"FF","rssi":-77,"servicedatauuid":"0x0d00","brand":"Switchbot","model":"S1","model_id":"SWITCHBOT-S1","mode":"on/off","state":"on","batt":91}
```

```json
{"id":"FF","rssi":-73,"servicedatauuid":"0x0d00","brand":"Switchbot","model":"S1","model_id":"SWITCHBOT-S1","mode":"onestate","state":"on","batt":91}
```

Regarding the state, I'm asking myself if we should have a boolean instead of on/off, just a food for thought.

**Reply by DigiH** [reply]
> Regarding the state, I'm asking myself if we should have a boolean instead of on/off, just a food for thought.

Then also renaming the key to "**on_state**" to make a **true/false** clearer? 

`"state":"on"` … `"state":"off"`
to
`"on_state":true` …`"on_state":false`

Personally, I prefer the on/off, clearer at first glance for me, but ultimately depends on how it can best be implemented into various backends.

I submitted a boolean version - slightly larger due to the pseudo function calls needed for **is_bool**, but assumedly a lot shorter when a direct bit-wise single property is possible - to see how it feels for you when using it with the device :)

**Reply by DigiH** [reply]
@1technophile Ok, I knew something seemed strange to me that for a boolean result a pseudo **value_from_hex_data** function call to the hex digit was needed, while the hex digit directly cannot really tell if the included bit, at whichever place, is '0' **false** or '1' **true**. That's why the multiple OR conditions are currently needed, to get **all** the hex digits, in this case Bit[6], to be evaluated as true or false.

Just testing with a new test case

"48005b"

which should definitely be

"servicedata", 2, "0" => in binary: '**0000**' => Bit[4] - Data Commit Flag 0: service data no data update / Bit[5] - Enc type 0: encryption algorithm (the same as the Bot) / Bit[6] - State 0: on / Bit[7] - Switchbot Mode 0: one state mode

so should be

`"{\"brand\":\"Switchbot\",\"model\":\"S1\",\"model_id\":\"SWITCHBOT-S1\",\"mode\":\"onestate\",\"on_state\":true,\"batt\":91}",`

but, likely with evaluating

```
"decoder":["value_from_hex_data", "servicedata", 2, 1, false],
"is_bool":1
```

hex 0 here as false, so the test run fails with

```
Expected: {"brand":"Switchbot","model":"S1","model_id":"SWITCHBOT-S1","mode":"onestate","on_state":true,"batt":91}
1: Got JSON: {"brand":"Switchbot","model":"S1","model_id":"SWITCHBOT-S1","mode":"onestate","on_state":false,"batt":91}
```

So to my understanding a true boolean can currently only be implemented if the state really corresponds to hex '0' always being **false** and hex '1' to 'f' always **true** - if no other bits in the hex' binary quadruplet influence it.

Or alternatively a **static_value** boolean state which can be implemented without any decoder call, without it just creating a string version "true"/"false" not really being boolean. Seems to be used like that in MUE4094RT_json.h but I couldn't find a test case, when trying it did however
```
Expected: {"brand":"Switchbot","model":"S1","model_id":"SWITCHBOT-S1","mode":"on/off","on_state":false,"batt":91}
Got JSON: {"brand":"Switchbot","model":"S1","model_id":"SWITCHBOT-S1","mode":"on/off","on_state":"false","batt":91}
```
so no real boolean advantage.

I'm not exactly sure what the circumstances are for other decoders which use a hex digit to avaluate a boolean state, possibly by the hex digit only being defined through one included true/false bit, but for this implementation it currently doesn't seem possible.

Or am I totally missing something?

Thanks

**Reply by 1technophile** [reply]
@DigiH I don't think you are missing something, let's go for on/off as this is more straight forward approach, sorry for wrong track :-(

**Reply by DigiH** [reply]
@1technophile back to on/off, at least for the time being

https://github.com/DigiH/decoder/tree/switchbot-s1

Could you confirm the info in the documentation, particularly the '1 CR2' battery info. Wasn't 100% sure if it's correct.

Thanks

**Reply by 1technophile** [reply]
I confirm the CR2!
